From 4be8b1b0a1e3be3efd58a880e81a29f53fe7cf54 Mon Sep 17 00:00:00 2001
From: Edward Hervey <edward@centricular.com>
Date: Tue, 3 May 2022 16:08:39 +0200
Subject: [PATCH 1/2] playbin3: Don't use unknown types for default selection

When creating a fallback default selection from a collection, don't attempt to
use unknown stream types

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/2353>
---
 gst/playback/gstdecodebin3.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gst/playback/gstdecodebin3.c b/gst-plugins-base/gst/playback/gstdecodebin3.c
index f8a2ec4bf64..d5cf44c8b27 100644
--- a/gst/playback/gstdecodebin3.c
+++ b/gst/playback/gstdecodebin3.c
@@ -1261,7 +1261,7 @@ update_requested_selection (GstDecodebin3 * dbin)
     for (i = 0; i < nb; i++) {
       GstStream *stream = gst_stream_collection_get_stream (collection, i);
       GstStreamType curtype = gst_stream_get_stream_type (stream);
-      if (!(used_types & curtype)) {
+      if (curtype != GST_STREAM_TYPE_UNKNOWN && !(used_types & curtype)) {
         const gchar *sid = gst_stream_get_stream_id (stream);
         GST_DEBUG_OBJECT (dbin,
             "Automatically selecting stream '%s' of type %s", sid,
-- 
GitLab


From 17aa70b8880659d2d30cb8a454bb2b89337bd847 Mon Sep 17 00:00:00 2001
From: Edward Hervey <edward@centricular.com>
Date: Tue, 3 May 2022 16:25:19 +0200
Subject: [PATCH 2/2] parsebin: Don't modify inexistant GstStream

When handling exposing un-handled streams, we can only replace the GstStream for
those we are creating ourselves (i.e. the fallback collection).

Fixes assertions when the demuxer creates those streams

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/2353>
---
 .../gst/playback/gstparsebin.c                | 27 ++++++++++---------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/gst/playback/gstparsebin.c b/gst-plugins-base/gst/playback/gstparsebin.c
index a8dca67fa33..d8480945295 100644
--- a/gst/playback/gstparsebin.c
+++ b/gst/playback/gstparsebin.c
@@ -1398,22 +1398,23 @@ analyze_new_pad (GstParseBin * parsebin, GstElement * src, GstPad * pad,
     g_value_array_free (factories);
 
     if (parsebin->expose_allstreams) {
-      GstStream *newstream;
-      GST_LOG_OBJECT (parsepad, "Existing GstStream %" GST_PTR_FORMAT,
-          parsepad->active_stream);
       /* If we expose all streams, we only need to inform the application about
        * a missing handler but still expose it. We also make sure the stream
        * type is unknown. */
-      g_assert (parsepad->active_stream);
-      newstream =
-          gst_stream_new (gst_stream_get_stream_id (parsepad->active_stream),
-          caps, GST_STREAM_TYPE_UNKNOWN,
-          gst_stream_get_stream_flags (parsepad->active_stream));
-      gst_object_replace ((GstObject **) & parsepad->active_stream,
-          (GstObject *) newstream);
-      GST_LOG_OBJECT (parsepad, "New GstStream %" GST_PTR_FORMAT,
-          parsepad->active_stream);
-
+      if (parsepad->in_a_fallback_collection) {
+        GstStream *newstream;
+        GST_LOG_OBJECT (parsepad, "Existing GstStream %" GST_PTR_FORMAT,
+            parsepad->active_stream);
+        g_assert (parsepad->active_stream);
+        newstream =
+            gst_stream_new (gst_stream_get_stream_id (parsepad->active_stream),
+            caps, GST_STREAM_TYPE_UNKNOWN,
+            gst_stream_get_stream_flags (parsepad->active_stream));
+        gst_object_replace ((GstObject **) & parsepad->active_stream,
+            (GstObject *) newstream);
+        GST_LOG_OBJECT (parsepad, "New GstStream %" GST_PTR_FORMAT,
+            parsepad->active_stream);
+      }
       gst_element_post_message (GST_ELEMENT_CAST (parsebin),
           gst_missing_decoder_message_new (GST_ELEMENT_CAST (parsebin), caps));
       goto expose_pad;
-- 
GitLab

